### ISSUE TITLE
Arbitrary data support

### ISSUE THREAD

**Issue by ericphanson** [dillondaudert/UMAP.jl#27]
It would be nice if UMAP.jl didn't require the input to be formatted as `X (a column-major matrix of shape (n_features, n_samples))`. For example, I have samples that I store as structs with some metadata and a data field, which is stored as a matrix itself. Then I define my distance (a subtype of `Distances.SemiMetric`) between these structs by grabbing their data fields and calculating a distance measure. It would be nice to be able to just pass this vector of samples and my distance to UMAP. In fact, NearestNeighborDescent already supports this (as long as you also define `Distances.result_type(::MyMetric, a, b) = Float32` or such), so I can get it working with UMAP just by removing some type annotations. It seems like UMAP only needs to know the number of original features to check that the output dimension is smaller, and they don't actually need to be formatted as a matrix.

(Of course, I could vectorize all the data matrices and `hcat` them into a big matrix, and then define my distance measure to reshape them, etc, but it seems unnecessary and complicated).

Does this seem like a reasonable feature for UMAP.jl?

**Comment by dillondaudert** [comment]
Yes, it's a very reasonable feature to have. You're right that UMAP.jl is unnecessarily type constrained, and I have plans in progress for refactoring it to be much more generic - one goal is to support datasets with arbitrary objects.

It should be ready at least to test out in a few weeks, I can ping you once it is if you like.

**Comment by ericphanson** [comment]
Sounds great! Thanks for the great package, by the way; I am new to dimensionality reduction but it was pretty easy to get some plots quickly with UMAP.jl and the NearestNeighborDescent.jl :).

**Comment by dillondaudert** [comment]
Just wanted to leave an update that I am still working on this, though it's taking me longer than I initially thought. Turns out the refactoring I had in mind is fairly substantial... 

If you're interested, you can take a look at the Wiki pages where there's a high level draft of where things are going.

**Comment by ericphanson** [comment]
I saw http://www.cs.cornell.edu/~arb/papers/GLANCE-WWW-2020.pdf today (from https://twitter.com/nassarhuda/status/1319675786085978114?s=21) and it seemed like an interesting way to visualize graphs. As I understand it, UMAP has three steps: calculate knns, construct a weighted graph, project it down. As you know, they have a specific reason for using the projection they do (trying to find a low dimensional structure with a similar topological structure) but it seems like it could still be interesting to try other methods too.

So just as I think it’s worthwhile to allow separate construction of the knns, I think it could be worthwhile to allow separate construction of the projection.
